yb-voyager/src/tgtdb/eventQueue.go

```diff
@@ -0,0 +1,144 @@
+limitations under the License.
+*/
+
+package tgtdb
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/yugabyte/yb-voyager/yb-voyager/src/utils"
+)
+
+type EventQueue struct {
+	QueueDirPath string
+}
+
+func NewEventQueue(exportDir string) *EventQueue {
+	return &EventQueue{
+		QueueDirPath: filepath.Join(exportDir, "data", "cdc"),
+	}
+}
+
+// GetNextSegments returns the next segments to process, in order.
+func (eq *EventQueue) GetNextSegments() ([]*EventQueueSegment, error) {
+	log.Infof("Getting next segments to process from %s", eq.QueueDirPath)
+	reSegmentName := regexp.MustCompile("queue.[0-9]+.ndjson$")
+	dirEntries, err := os.ReadDir(eq.QueueDirPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read dir %s: %w", eq.QueueDirPath, err)
+	}
+	var segmentPaths []string
+	for _, dirEntry := range dirEntries {
+		if !dirEntry.IsDir() && reSegmentName.MatchString(dirEntry.Name()) {
+			segmentPaths = append(segmentPaths, filepath.Join(eq.QueueDirPath, dirEntry.Name()))
+		}
+	}
+
+	segments := make([]*EventQueueSegment, 0, len(segmentPaths))
+	for _, segmentPath := range segmentPaths {
+		segmentFileName := filepath.Base(segmentPath)
+		var segmentNum int64
+		_, err := fmt.Sscanf(segmentFileName, "queue.%d.ndjson", &segmentNum)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse segment number from %s: %w", segmentFileName, err)
+		}
+		segments = append(segments, NewEventQueueSegment(segmentPath, segmentNum))
+	}
+
+	sort.Slice(segments, func(i, j int) bool {
+		return segments[i].SegmentNum < segments[j].SegmentNum
+	})
+
+	return segments, nil
+}
+
+type EventQueueSegment struct {
+	FilePath   string
+	SegmentNum int64 // 0-based
+	processed  bool
+	file       *os.File
+	reader     *utils.TailReader
+}
+
+var EOFMarker = `\.`
+
+func NewEventQueueSegment(filePath string, segmentNum int64) *EventQueueSegment {
+	return &EventQueueSegment{
+		FilePath:   filePath,
+		SegmentNum: segmentNum,
+		processed:  false,
+	}
+}
+
+func (eqs *EventQueueSegment) Open() error {
+	file, err := os.OpenFile(eqs.FilePath, os.O_RDONLY, 0640)
+	if err != nil {
+		return fmt.Errorf("failed to open segment file %s: %w", eqs.FilePath, err)
+	}
+	eqs.file = file
+	eqs.reader = utils.NewTailReader(file)
+	return nil
+}
+
+func (eqs *EventQueueSegment) Close() error {
+	return eqs.file.Close()
+}
+
+// ReadEvent reads an event from the segment file.
+// Waits until an event is available.
+func (eqs *EventQueueSegment) NextEvent() (*Event, error) {
+	var event Event
+	line, err := eqs.reader.ReadLine()
+	if err != nil && err != io.EOF {
+		return nil, fmt.Errorf("failed to read line from %s: %w", eqs.FilePath, err)
+	}
+
+	if string(line) == EOFMarker && err == io.EOF {
+		log.Infof("reached EOF marker in segment %s", eqs.FilePath)
+		eqs.processed = true
+		return nil, nil
+	}
+
+	err = json.Unmarshal(line, &event)
```

Comment:
Is this unmarshal on a best-effort basis or does it ensure that all fields of the `Event` struct are populated? 

Just want to avoid scenarios where the all fields of the json are not unmarshalled, thereby leading to the event's fields having default values (for whatever reason)

Reply:
In case of JSON unmarshal in golang, either the default is provided by the struct or it will be taken as empty string.